% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/pileupMT.R
\name{pileupMT}
\alias{pileupMT}
\title{Pileup the mitochondrial reads in a BAM, for variant calling or CN purposes.}
\usage{
pileupMT(bam, sbp = NULL, pup = NULL, parallel = FALSE, cores = 1,
  ref = c("rCRS", "GRCh37", "GRCh38", "hg38", "GRCm38", "C57BL/6J",
  "NC_005089", "mm10"), ...)
}
\arguments{
\item{bam}{BAM (must be indexed!) file name or object with a @bam slot}

\item{sbp}{optional ScanBamParam object (autogenerated if missing)}

\item{pup}{optional PileupParam object (autogenerated if missing)}

\item{parallel}{set to TRUE to run in parallel}

\item{cores}{if unassigned and parallel, uses half of available cores}

\item{ref}{aligned reference mitogenome (default is rCRS/GRCh37+)}

\item{...}{additional args to pass on to the pileup() function}
}
\value{
an MVRanges object, constructed from pileup results
}
\description{
If a BAM filename is given, but no ScanBamParam, scanMT(bam) will be called. 
Human mitochondrial genomes (GRCh37+ and hg38+) are fully supported, mouse
mitochondrial genomes (C57BL/6J aka NC_005089) are only somewhat supported.
}
\examples{

library(MTseekerData)
BAMdir <- system.file("extdata", "BAMs", package="MTseekerData")

bam <- BAMs[1]

sbp <- scanMT(bam)
show(sbp)

pu <- pileupMT(bam, ref="rCRS", sbp=sbp)
show(pu) 

}
