#' Pileup the mitochondrial reads in a BAM, for variant calling or CN purposes. 
#'
#' If a BAM filename is given, but no ScanBamParam, scanMT(bam) will be called. 
#' Human mitochondrial genomes (GRCh37+ and hg38+) are fully supported, mouse
#' mitochondrial genomes (C57BL/6J aka NC_005089) are only somewhat supported.
#' 
#' @param bam       BAM (must be indexed!) file name or object with a @bam slot
#' @param sbp       optional ScanBamParam object (autogenerated if missing) 
#' @param pup       optional PileupParam object (autogenerated if missing)
#' @param ref       aligned reference mitogenome (default is rCRS/GRCh37+)
#' @param ...       additional args to pass on to the pileup() function
#'
#' @return          an MVRanges object, constructed from pileup results 
#' 
#' @import GenomicAlignments
#' @import GenomeInfoDb
#' @import Rsamtools
#' 
#' @examples
#' 
#' library(MTseekerData)
#' BAMdir <- system.file("extdata", "BAMs", package="MTseekerData")
#' BAMs <- paste0(BAMdir, "/", list.files(BAMdir, pattern=".bam$"))
#' bam <- BAMs[1]
#' 
#' sbp <- scanMT(bam)
#' show(sbp)
#' 
#' pu <- pileupMT(bam, ref="rCRS", sbp=sbp)
#' show(pu) 
#'
#' @export
#'
pileupMT <- function(bam, sbp=NULL, pup=NULL, ref=c("rCRS","GRCh37","GRCh38","hg38", "GRCm38","C57BL/6J","NC_005089","mm10"), ...) { 
  
  # can support multiple bams if we have sample names
  # perhaps it is worthwhile to autoextract them now
  
  if (is.null(sbp)) sbp <- scanMT(bam) 
  if (is.null(pup)) pup <- PileupParam(distinguish_strands=FALSE) 
  refSeqLengths <- .refSeqLengths() # synonyms 
  ref <- .getRefSyn(match.arg(ref)) # matching 
  
  # scant support for other mitogenomes
  if (!ref %in% names(refSeqLengths)) {
    stop("Only the ", paste(names(refSeqLengths), collapse=" and "),
         " mitochondrial genome", 
         ifelse(length(refSeqLengths) > 1, "s are", " is"),
         " supported by pileupMT() at present.")
  } else { 
    pu <- pileup(file=bam, scanBamParam=sbp, pileupParam=pup, ...)
  }
  
  # may be handy for editing 
  refSeqDNA <- .getRefSeq(ref)
  
  # will need to handle '-' and '+' separately 
  indels <- subset(pu, nucleotide %in% c('-', '+'))
  if (nrow(indels) > 0) {
    # for obtaining the supporting reads and CIGARs from the BAM: 
    indels$start <- indels$pos
    indels$end <- indels$pos 
    indelSBP <- sbp
    bamWhich(indelSBP) <- as(indels, "GRanges")
    indelReads <- readGAlignments(file=bam, param=indelSBP)
    
    # Get the alternative sequences
    mcols(indelReads)$indelStart <- NA_integer_
    mcols(indelReads)$indelEnd <- NA_integer_
    mcols(indelReads)$ref <- NA_character_
    mcols(indelReads)$alt <- NA_character_
    
    for (i in 1:length(indelReads)) {
      indelReads[i] <- .reverseCigar(indelReads[i], ref)
    }
    
    mvrIndel <- .indelToMVR(indelReads, ref)
  
    # Copied and pasted from below
    mvrIndel$VAF <- altDepth(mvrIndel)/totalDepth(mvrIndel)
    metadata(mvrIndel)$refseq <- .getRefSeq(ref)
    covg <- rep(0, length(metadata(mvrIndel)$refseq))
    
    ### Not sure how to apply this to the indels
    #covered <- rowsum(pu$count, pu$pos)
    #covg[as.numeric(rownames(covered))] <- covered 
    #metadata(mvr)$coverageRle <- Rle(covg)
    
    metadata(mvrIndel)$bam <- basename(bam)
    metadata(mvrIndel)$sbp <- sbp
    metadata(mvrIndel)$pup <- pup
    mvrIndel$bam <- basename(bam)
    genome(mvrIndel) <- ref
    
    # cigar(indelReads)
    # 
    
    
    message("Warning: indels are not currently supported in pileupMT()")
    message("(However, if you debug() this function, indelReads will help.")
  }
  # data(fpFilter_Triska, package="MTseeker") # for when they are... 
  
  # this may belong in a separate helper function...
  pu <- subset(pu, nucleotide %in% c('A','C','G','T'))
  pu$which_label <- NULL # confusing here 
  pu$ref <-  factor(strsplit(as.character(.getRefSeq(ref)), '')[[1]],
                    levels=levels(pu$nucleotide))[pu$pos]
  
  pu$alt <- pu$nucleotide 
  pu$totalDepth <- .byPos(pu, "count")
  is.na(pu$alt) <- (pu$nucleotide == pu$ref)
  pu$altDepth <- ifelse(is.na(pu$alt), NA_integer_, pu$count) 
  pu$refDepth <- pu$totalDepth - .byPos(pu, "altDepth")
  pu$isAlt <- !is.na(pu$alt)
  pu$alleles <- .byPos(pu, "isAlt") + 1 
  columns <- c("seqnames","pos","ref","alt","totalDepth","refDepth","altDepth")
  gr <- keepSeqlevels(.puToGR(subset(pu, isAlt|alleles==1)[,columns]),
                      unique(pu$seqnames))
  seqinfo(gr) <- Seqinfo("chrM", width(.getRefSeq(ref)), 
                         isCircular=TRUE, genome=ref)
  vr <- makeVRangesFromGRanges(.puToGR(subset(pu, isAlt|alleles==1)[,columns]))
  vr <- keepSeqlevels(vr, "chrM") 
  seqinfo(vr) <- seqinfo(gr)
  mvr <- MVRanges(vr, coverage=median(rowsum(pu$count, pu$pos)))
  sampleNames(mvr) <- base::sub(paste0(".", ref), "", 
                                base::sub(".bam", "", 
                                          basename(bam)))
  names(mvr)[which(mvr$ref != mvr$alt)] <- MTHGVS(subset(mvr, ref != alt)) 
  altDepth(mvr)[is.na(altDepth(mvr))] <- 0
  
  mvr$VAF <- altDepth(mvr)/totalDepth(mvr)
  metadata(mvr)$refseq <- .getRefSeq(ref)
  covg <- rep(0, length(metadata(mvr)$refseq))
  covered <- rowsum(pu$count, pu$pos)
  covg[as.numeric(rownames(covered))] <- covered 
  metadata(mvr)$coverageRle <- Rle(covg)
  metadata(mvr)$bam <- basename(bam)
  metadata(mvr)$sbp <- sbp
  metadata(mvr)$pup <- pup
  mvr$bam <- basename(bam)
  genome(mvr) <- ref
  
  # Only want to keep variants that differ from reference
  keep <- which(!is.na(alt(mvr)))
  mvr <- mvr[keep]
  names(mvr) <- MTHGVS(mvr)
  
  # Add in the indels
  mvr <- MVRanges(c(mvr, mvrIndel))
  mvr <- sort(mvr)
  
  return(mvr)
}


# helper fn
.byPos <- function(x, y) { 
  rowsum(as.numeric(x[,y]), as.character(x$pos), na.rm=1)[as.character(x$pos),]
}


# helper fn
.puToGR <- function(pu) { 
  makeGRangesFromDataFrame(pu, start.field="pos", end.field="pos", strand="*",
                           keep.extra.columns=TRUE)
}


# helper fn
.refSeqLengths <- function() { 
  
  refSeqLengths <- c() 
  data(rCRSeq, package="MTseeker") # human
  data(NC_005089seq, package="MTseeker") # mouse 
  # data(mtRefs, package="MTseeker") # both
  refSeqLengths["rCRS"] <- width(rCRSeq)
  refSeqLengths["GRCh38"] <- width(rCRSeq)
  refSeqLengths["hg38"] <- width(rCRSeq)
  refSeqLengths["GRCh37"] <- width(rCRSeq)
  refSeqLengths["NC_005089"] <- width(NC_005089seq)
  refSeqLengths["GRCm38"] <- width(NC_005089seq)
  refSeqLengths["mm10"] <- width(NC_005089seq) 
  refSeqLengths["C57BL/6J"] <- width(NC_005089seq)
  return(refSeqLengths)
  
}


# helper fn
.getRefSyn <- function(ref) {
  
  # simplify references down to the supported "mouse" or "human" mitogenomes 
  if (ref %in% c("C57BL/6J", "NC_005089", "GRCm38","mm10")) ref <- "NC_005089"
  if (ref %in% c("hg38", "GRCh37", "GRCh38")) ref <- "rCRS" 
  return(ref)
  
}


# helper fn
.getRefSeq <- function(ref) { 
  
  data(rCRSeq, package="MTseeker") # human
  data(NC_005089seq, package="MTseeker") # mouse
  refs <- DNAStringSet(c(rCRS=rCRSeq[['chrM']],
                         NC_005089=NC_005089seq[['chrM']]))
  return(refs[.getRefSyn(ref)])
  
}


.reverseCigar <- function(indelRead, ref) {

  # These are reads that support the reference
  if ( (!grepl("I", cigar(indelRead))) && (!grepl("D", cigar(indelRead))) ) {
    return(indelRead)
    #print("All matched")
  }
  
  ### HOW AM I SUPPOSED TO KNOW WHAT WAS INSERTED
  if (grepl("I", cigar(indelRead))) {
    warning("NEED A FIX FOR INSERTIONS ", cigar(indelRead))
    return(indelRead)
    #indelRead$ref <- "insertion"
  }
  
  else {
    addPos <- 0
    
    reference <- .getRefSeq(ref)
    
    splitCigar <- gsub("([[:digit:]]+[[:alpha:]])", "\\1 ", cigar(indelRead))
    splitCigar <- unlist(strsplit(splitCigar, " "))
    
    for (i in 1:length(splitCigar)) {
      
      # Soft clippings are ignored
      if (grepl("S", splitCigar[i])) next
      
      # Matched reads
      if (grepl("M", splitCigar[i])) {
        addPos <- addPos + as.numeric(gsub("\\D", "", splitCigar[i]))
      } # M
      
      # Deletions
      if (grepl("D", splitCigar[i])) {
        
        # Create range for the deletion
        # When comparing against the old way of doing things, it seems to always be off by 1
        startPos <- start(indelRead) + addPos - 1
        endPos <- startPos + as.numeric(gsub("\\D", "", splitCigar[i]))
        
        mcols(indelRead)$indelStart <- startPos
        mcols(indelRead)$indelEnd <- endPos
        
        # Extract reference sequence
        ref <- extractAt(reference, IRanges(startPos, endPos))
        ref <- unstrsplit(CharacterList(ref))
        mcols(indelRead)$ref <- unname(ref)
        
        # Do the deletion
        # Since we are subtracting 1, we are deleting all of the sequences that come after the start
        # So we just want to keep the first sequence
        alt <- extractAt(reference, IRanges(startPos, startPos))
        alt <- unstrsplit(CharacterList(alt))
        mcols(indelRead)$alt <- unname(alt)
        
      } # D
    } # for
  } # Else
  return(indelRead)
}


.indelToMVR <- function(indelReads, ref) {
  
  # This is mostly copied and pasted from the mvr created at the end of pileup
  indelSupport <- which(!is.na(mcols(indelReads)$alt))
  indelSupport <- indelReads[indelSupport]
  indelSupport <- as.data.frame(indelSupport)
  
  indelSupport$which_label <- NULL
  
  indelSupport$refDepth <- NA_integer_
  indelSupport$altDepth <- 1
  indelSupport$totalDepth <- 1
  
  indelSupport$isAlt <- TRUE
  
  ### Not sure what this line is supposed to do
  #pu$alleles <- .byPos(pu, "isAlt") + 1 
  
  #### DO I WANT TO DO THIS??
  indelSupport$pos <- as.numeric(indelSupport$indelStart)
  
  # Turn into granges then turn into vranges
  columns <- c("seqnames","pos","ref","alt","totalDepth","refDepth","altDepth")
  grIndel <- keepSeqlevels(.puToGR(subset(indelSupport, isAlt==1)[,columns]), unique(indelSupport$seqnames))
  seqinfo(grIndel) <- Seqinfo("chrM", width(.getRefSeq(ref)), isCircular=TRUE, genome=ref)
  
  vrIndel <- makeVRangesFromGRanges(.puToGR(subset(indelSupport, isAlt==1)[,columns]))
  vrIndel <- keepSeqlevels(vrIndel, "chrM") 
  seqinfo(vrIndel) <- seqinfo(grIndel)
  
  # Skipped this coverage step
  ### IS THAT ALRIGHT?
  #mvr <- MVRanges(vr, coverage=median(rowsum(pu$count, pu$pos)))
  
  # Turn into mvranges
  mvrIndel <- MVRanges(vrIndel)
  sampleNames(mvrIndel) <- base::sub(paste0(".", ref), "", base::sub(".bam", "",  basename(bam)))
  
  # Fix ranges
  indelRanges <- IRanges(as.numeric(indelSupport$indelStart), as.numeric(indelSupport$indelEnd))
  ranges(mvrIndel) <- indelRanges
  
  # Assign names
  names(mvrIndel)[which(ref(mvrIndel) != alt(mvrIndel))] <- MTHGVS(subset(mvrIndel, ref != alt)) 
  
  # Only keep unique variants (based upon names that were just assigned)
  mvrIndelunique <- unique(mvrIndel)
  
  # Assign altDepths based upon what was unique
  # Sometimes the order gets mixed up, so gotta keep track of that
  order <- match(names(mvrIndelunique), names(table(names(mvrIndel))))
  count <- unname(table(names(mvrIndel))[order])
  altDepth(mvrIndelunique) <- count
  
  # Now have to get the refDepth from indelReads that were all matched
  #refDepth(mvrIndelunique) <- 0
  refSupport <- which(is.na(mcols(indelReads)$alt))
  refSupport <- indelReads[refSupport]
  
  # Need to keep order in mind, since not every indel has a read supporting the reference
  ov <- findOverlaps(refSupport, mvrIndelunique)
  refCount <- table(subjectHits(ov))
  refDepth(mvrIndelunique[as.numeric(names(refCount))]) <- unname(refCount)
  
  # Replace NA's with 0s for refDepth to properly calculate totalDepth
  noRef <- which(is.na(refDepth(mvrIndelunique)))
  refDepth(mvrIndelunique[noRef]) <- 0
  
  # Recalculate totalDepth
  totalDepth(mvrIndelunique) <- refDepth(mvrIndelunique) + altDepth(mvrIndelunique)
  
  return(mvrIndelunique)
  
}
