#' Pileup the mitochondrial reads in a BAM, for variant calling or CN purposes. 
#'
#' If a BAM filename is given, but no ScanBamParam, scanMT(bam) will be called. 
#' Human mitochondrial genomes (GRCh37+ and hg38+) are fully supported, mouse
#' mitochondrial genomes (C57BL/6J aka NC_005089) are only somewhat supported.
#' 
#' @param bam       BAM (must be indexed!) file name or object with a @bam slot
#' @param sbp       optional ScanBamParam object (autogenerated if missing) 
#' @param pup       optional PileupParam object (autogenerated if missing)
#' @param parallel  set to TRUE to run in parallel
#' @param cores     if unassigned and parallel, uses half of available cores
#' @param ref       aligned reference mitogenome (default is rCRS/GRCh37+)
#' @param ...       additional args to pass on to the pileup() function
#'
#' @return          an MVRanges object, constructed from pileup results 
#' 
#' @import GenomicAlignments
#' @import GenomeInfoDb
#' @import Rsamtools
#' 
#' @examples
#' 
#' library(MTseekerData)
#' BAMdir <- system.file("extdata", "BAMs", package="MTseekerData")
#' 
#' bam <- BAMs[1]
#' 
#' sbp <- scanMT(bam)
#' show(sbp)
#' 
#' pu <- pileupMT(bam, ref="rCRS", sbp=sbp)
#' show(pu) 
#'
#' @export
#'
pileupMT <- function(bam, sbp=NULL, pup=NULL, parallel=FALSE, cores=1, ref=c("rCRS","GRCh37","GRCh38","hg38", "GRCm38","C57BL/6J","NC_005089","mm10"), ...) { 
  
  # List of reference genomes and their lengths
  refSeqLengths <- .refSeqLengths() # synonyms 
  
  # Stop if the user forget to specify a reference genome
  if (length(ref) > 1) {
    stop("Forgot to set a reference genome! MTseeker currently supports: ", paste(names(refSeqLengths), collapse= ", "))
  }
  
  # scant support for other mitogenomes
  if (!ref %in% names(refSeqLengths)) {
    stop("Only the ", paste(names(refSeqLengths), collapse=" and "),
         " mitochondrial genome", 
         ifelse(length(refSeqLengths) > 1, "s are", " is"),
         " supported by pileupMT() at present.")
  }
  
  # Determine which reference genome to use
  # Currently supports mouse and human
  ref <- .getRefSyn(match.arg(ref)) # matching 
  
  # Set the number of cores if running in parallel
  if (parallel) options(mc.cores = cores)
  if (parallel & cores == 1) options(mc.cores = detectCores()/2)
  
  # If you are working with multiple files
  if (length(bam) > 1) {
    
    if(parallel) {
      mvrl <- MVRangesList(mclapply(bam, pileupMT, sbp=sbp, ref=ref, parallel=parallel, cores=cores))
      
      ### THIS WILL BREKAK IF YOU FEED IT EMPTY MVRANGES
      ### FIX LATER
      sampNames <- unlist(lapply(mvrl, function(x) unique(as.character(sampleNames(x)))))
      names(mvrl) <- sampNames
      
      return(mvrl)
    }
    
    else {
      mvrl <- MVRangesList(lapply(bam, pileupMT, ref=ref, sbp=sbp))
      
      ### THIS WILL BREKAK IF YOU FEED IT EMPTY MVRANGES
      ### FIX LATER
      sampNames <- unlist(lapply(mvrl, function(x) unique(as.character(sampleNames(x)))))
      names(mvrl) <- sampNames
      
      return(mvrl)
    }
  }
  
  # can support multiple bams if we have sample names
  # perhaps it is worthwhile to autoextract them now
  
  if (is.null(sbp)) sbp <- scanMT(bam) 
  if (is.null(pup)) pup <- PileupParam(distinguish_strands=FALSE) 
  
  pu <- pileup(file=bam, scanBamParam=sbp, pileupParam=pup, ...)
  
  # may be handy for editing 
  refSeqDNA <- .getRefSeq(ref)
  
  # Need the total number of reads in order to calculate coverage
  readsSBP <- sbp
  bamWhat(readsSBP) <- "seq"
  reads <- readGAlignments(file=bam, param=readsSBP)
  numReads <- length(reads)
  readsWidth <- median(qwidth(reads)) - 1
  
  # number of reads * length of reads / length of ref sequence
  covg <- round(numReads * readsWidth / width(refSeqDNA))
  if (is.na(covg)) covg <- 0
  
  # will need to handle '-' and '+' separately 
  indels <- subset(pu, nucleotide %in% c('-', '+'))
  if (nrow(indels) > 0) {
    # for obtaining the supporting reads and CIGARs from the BAM: 
    indels$start <- indels$pos
    indels$end <- indels$pos 
    indelSBP <- sbp
    bamWhich(indelSBP) <- as(indels, "GRanges")
    bamWhat(indelSBP) <- "seq"
    indelReads <- readGAlignments(file=bam, param=indelSBP)
    
    # Initialize to get the alternative sequences
    mcols(indelReads)$indelStart <- NA_integer_
    mcols(indelReads)$indelEnd <- NA_integer_
    
    mcols(indelReads)$ref <- NA_character_
    mcols(indelReads)$alt <- NA_character_
    
    # Get the reads the physically contain the insertion and deletions
    indelIndex <- which(grepl("D|I", cigar(indelReads)))
    
    # Every other read will support the reference
    everything <- seq(1, length(indelReads), 1)
    refIndex <- which(is.na(match(everything, indelIndex)))
    refSupport <- indelReads[refIndex]
    
    # Now this contains just indel reads
    # Hopefully this will speed things up with less iteration
    indelReads <- indelReads[indelIndex]
    
    message("Tallying indels for: ", bam)

    if (length(indelIndex) > 500) {
      message("This may take a while, determining ", length(indelIndex), " reads for indels from ", bam)
    }
    
    # Create new data structure
    # Have to do this because sometimes an indelRead will have multiple indels within the read
    newIndelReads <- indelReads[0]

    # Returns the ref and alt read for each variant
    #lapply(indelReads, .reverseCigar, ref=ref, reference=refSeqDNA)
    for (i in 1:length(indelReads)) {
      newIndelReads <- append(newIndelReads, .reverseCigar(indelReads[i], ref, refSeqDNA, i))
    }
    
    # Converts to a MVRanges
    mvrIndel <- .indelToMVR(newIndelReads,refSupport, ref, bam, covg) 

    # Copied and pasted from below
    mvrIndel$VAF <- altDepth(mvrIndel)/totalDepth(mvrIndel)
    metadata(mvrIndel)$refseq <- refSeqDNA
    
    metadata(mvrIndel)$bam <- basename(bam)
    metadata(mvrIndel)$sbp <- indelSBP
    metadata(mvrIndel)$pup <- pup
    mvrIndel$bam <- basename(bam)
    genome(mvrIndel) <- ref
    
  }
  # data(fpFilter_Triska, package="MTseeker") # for when they are... 
  
  message("Tallying SNPs for: ", bam)

  # this may belong in a separate helper function...
  pu <- subset(pu, nucleotide %in% c('A','C','G','T'))
  pu$which_label <- NULL # confusing here 
  pu$ref <-  factor(strsplit(as.character(refSeqDNA), '')[[1]],
                    levels=levels(pu$nucleotide))[pu$pos]
  
  pu$alt <- pu$nucleotide 
  pu$totalDepth <- .byPos(pu, "count")
  is.na(pu$alt) <- (pu$nucleotide == pu$ref)

  # Only want to keep variants that differ from reference
  keep <- which(!is.na(pu$alt))
  pu <- pu[keep,]
  
  # There are no indels and no SNPs
  if (nrow(pu) == 0 && nrow(indels) == 0) {
    return(MVRanges(GRanges(c(seqnames=NULL,ranges=NULL,strand=NULL)), coverage = covg))
  }
  
  # If there are no SNPs but there are indels
  else if (nrow(pu) == 0 && nrow(indels) > 0 ) {
    metadata(mvrIndel)$coverageRle <- coverage(mvrIndel)
    return(mvrIndel)
  }
  
  # If there exists SNPs 
  if (nrow(pu) != 0) {
   
    pu$altDepth <- ifelse(is.na(pu$alt), NA_integer_, pu$count) 
    pu$refDepth <- pu$totalDepth - .byPos(pu, "altDepth")
    pu$isAlt <- !is.na(pu$alt)
    pu$alleles <- .byPos(pu, "isAlt") + 1 
    columns <- c("seqnames","pos","ref","alt","totalDepth","refDepth","altDepth")
    gr <- keepSeqlevels(.puToGR(subset(pu, isAlt|alleles==1)[,columns]),
                        unique(pu$seqnames))
    
    # seqnames for mouse genome is "MT" 
    seqinfo(gr) <- Seqinfo(levels(seqnames(gr)), width(refSeqDNA), 
                           isCircular=TRUE, genome=ref)
    vr <- makeVRangesFromGRanges(.puToGR(subset(pu, isAlt|alleles==1)[,columns]))
    vr <- keepSeqlevels(vr, levels(seqnames(gr))) 
    seqinfo(vr) <- seqinfo(gr)
    #mvr <- MVRanges(vr, coverage=median(rowsum(pu$count, pu$pos)))
    mvr <- MVRanges(vr, coverage=covg)
    sampleNames(mvr) <- base::sub(paste0(".", ref), "", 
                                  base::sub(".bam", "", 
                                            basename(bam)))
    names(mvr)[which(mvr$ref != mvr$alt)] <- MTHGVS(subset(mvr, ref != alt)) 
    altDepth(mvr)[is.na(altDepth(mvr))] <- 0

    mvr$VAF <- altDepth(mvr)/totalDepth(mvr)
    metadata(mvr)$refseq <- refSeqDNA

    metadata(mvr)$bam <- basename(bam)
    metadata(mvr)$sbp <- sbp
    metadata(mvr)$pup <- pup
    mvr$bam <- basename(bam)
    genome(mvr) <- ref
     
    names(mvr) <- MTHGVS(mvr)
  }
  
  # Merge indels and SNPs together if both exist
  if (nrow(indels) > 0 && nrow(pu) != 0) {
    
    mvr <- MVRanges(c(mvr, mvrIndel), coverage=covg)
    mvr <- sort(mvr)
    names(mvr) <- MTHGVS(mvr)
  }

  metadata(mvr)$coverageRle <- coverage(mvr)
  return(mvr)
}


# helper fn
.byPos <- function(x, y) { 
  rowsum(as.numeric(x[,y]), as.character(x$pos), na.rm=1)[as.character(x$pos),]
}


# helper fn
.puToGR <- function(pu) { 
  makeGRangesFromDataFrame(pu, start.field="pos", end.field="pos", strand="*",
                           keep.extra.columns=TRUE)
}


# helper fn
.refSeqLengths <- function() { 
  
  refSeqLengths <- c() 
  data(rCRSeq, package="MTseeker") # human
  data(NC_005089seq, package="MTseeker") # mouse 
  # data(mtRefs, package="MTseeker") # both
  refSeqLengths["rCRS"] <- width(rCRSeq)
  refSeqLengths["GRCh38"] <- width(rCRSeq)
  refSeqLengths["hg38"] <- width(rCRSeq)
  refSeqLengths["GRCh37"] <- width(rCRSeq)
  refSeqLengths["NC_005089"] <- width(NC_005089seq)
  refSeqLengths["GRCm38"] <- width(NC_005089seq)
  refSeqLengths["mm10"] <- width(NC_005089seq) 
  refSeqLengths["C57BL/6J"] <- width(NC_005089seq)
  return(refSeqLengths)
  
}


# helper fn
.getRefSyn <- function(ref) {
  
  # simplify references down to the supported "mouse" or "human" mitogenomes 
  if (ref %in% c("C57BL/6J", "NC_005089", "GRCm38","mm10")) ref <- "NC_005089"
  if (ref %in% c("hg38", "GRCh37", "GRCh38")) ref <- "rCRS" 
  return(ref)
  
}


# helper fn
.getRefSeq <- function(ref) { 
  
  data(rCRSeq, package="MTseeker") # human
  data(NC_005089seq, package="MTseeker") # mouse
  refs <- DNAStringSet(c(rCRS=rCRSeq[['chrM']],
                         NC_005089=NC_005089seq[['chrM']]))
  return(refs[.getRefSyn(ref)])
  
}

# helper fn
# Will take a single read and return the reference and alternative sequence
#.reverseCigar <- function(indelRead, ref, reference) {
.reverseCigar <- function(indelRead, ref, reference, index) {

  # These are reads that support the reference (no indels found in them)
  if ( (!grepl("I", cigar(indelRead))) && (!grepl("D", cigar(indelRead))) ) {
    return(indelRead[0])
  }
  
  else {
    
    # Keep track of how many base pairs are being added in the cigar string
    addPos <- 0 # Matches (M in cigar strings)
    softPos <- 0 # Soft clips (S in cigar strings)
    
    splitCigar <- gsub("([[:digit:]]+[[:alpha:]])", "\\1 ", cigar(indelRead))
    splitCigar <- unlist(strsplit(splitCigar, " "))
    
    # Find which elements of the cigar string are insertion or deletion
    # Are there multiple insertions and deletions?
    indelIndex <- which((grepl("I|D", splitCigar))) 
    
    # Create a new indel read to edit
    newIndelRead <- indelRead

    # Iterate through in case there are multiple deletions and insertions
    for (i in 1:length(indelIndex)) {

      # Find the number of base pairs that are soft clips
      # Only want to look at the information before the currently considered indel
      softPos <- grep("S", splitCigar[1:indelIndex[i]])
      softPos <- sum(as.numeric(gsub("\\D", "", splitCigar[softPos])))
      if (length(softPos) == 0) softPos <- 0
      
      # Find the number of base pairs that are matches
      # Only want to look at the information before the currently considered indel
      addPos <- grep("M", splitCigar[1:indelIndex[i]])
      addPos <- sum(as.numeric(gsub("\\D", "", splitCigar[addPos])))
      if (length(addPos) == 0) addPos <- 0
      
      # If there is a previous indel in the same read
      # Need to consider how the start position will change on the reference
      # If there was an insertion beforehand, be careful when calculation startPos
      # Because those bp are adding length that does not exist into the ref seq
      
      # Add when there is a deletion
      # Subtract when there is an insertion
      if (i != 1) {
        #browser()
        # Indices or previous insertions or deletions in the same read
        prevIns <- which(grepl("I", splitCigar[1: (indelIndex[i] - 1) ]))
        prevDel <- which(grepl("D", splitCigar[1: (indelIndex[i] - 1) ]))
        
        # Sums of how many bp are deleted or inserted
        prevInsSum <- sum(as.numeric(gsub("\\D", "", splitCigar[prevIns])))
        prevDelSum <- sum(as.numeric(gsub("\\D", "", splitCigar[prevDel]))) 
        
      } 

      ## Now do the actual insertions or deletions

      # Insertions
      if (grepl("I", splitCigar[indelIndex[i]])) {
        
        # Start of range for the reference
        # Always off by one 
        startPos <- start(indelRead) + addPos - 1
        
        # Get the start index for insertion for the raw read
        altIndexStart <- addPos + softPos 
        
        # Multiple indels
        # Must take into consideration how that will effect the start positions
        # Add to the reference with previous deletion (insertions do not exist in the reference)
        # Add to the alternate with previous insertion (deletions do not add length to the alternate)
        if (i != 1) {
          #browser()
          startPos <- startPos + prevDelSum
          altIndexStart <- altIndexStart + prevInsSum
        }
        
        # According to VRanges documentation, insertions have width=1
        # So start and end are the same for the reference
        endPos <- startPos
        
        # Get the reference
        refs <- extractAt(reference, IRanges(startPos, endPos))
        refs <- unname(as.character(unlist(refs)))
        
        # Length of insertion
        insLength <- as.numeric(gsub("\\D", "", splitCigar[indelIndex[i]]))
        
        # Get the alt sequence from the raw read
        altSeq <- as.character(mcols(indelRead)$seq)
        altSplit <- unlist(strsplit(altSeq, split=""))
        altIndex <- seq(altIndexStart, altIndexStart + insLength, 1)
        altInd <- altSplit[altIndex]
        alt <- paste(altInd, collapse="")
        
        # A simple test to make sure insertion was done correctly
        if (altInd[1] != refs && length(indelIndex) == 1) {
          message("Incorrect indel for insertion from indel read: ", index)
          comp <- .sanCheck(reference, startPos, indelRead, softPos, refs, alt, splitCigar)
        }
        
        # Sanity check!
        #comp <- .sanCheck(reference, startPos, indelRead, softPos, refs, alt, splitCigar)

      } # I
      
      # Deletions
      else if (grepl("D", splitCigar[indelIndex[i]])) {

        # Create range for the deletion
        # This must be zero indexed or something, because it is always off by one
        startPos <- start(indelRead) + addPos - 1
        
        # Get the index for where the deletion takes place
        altIndexStart <- addPos + softPos 
        
        # Multiple indels
        # Must take into consideration how the start and end values will change
        if (i != 1) {
          #browser()
          startPos <- startPos + prevDelSum
          altIndexStart <- altIndexStart + prevInsSum
        }
        
        # Length of deletion and end position
        delLength <- as.numeric(gsub("\\D", "", splitCigar[indelIndex[i]]))
        endPos <- startPos + delLength
        
        # Get the reference
        refs <- extractAt(reference, IRanges(startPos, endPos))
        refs <- unname(as.character(unlist(refs)))
        refSplit <- unlist(strsplit(refs, split=""))
        
        # Get the alt sequence from the raw read
        altSeq <- as.character(mcols(indelRead)$seq)
        altSplit <- unlist(strsplit(altSeq, split=""))
        alt <- altSplit[altIndexStart]

        # A simple check to ensure the deletion is supposedly happening correctly
        if ( (alt != refSplit[1]) && length(indelIndex) == 1) {
          message("Incorrect deletion for indel read, ", index)
          comp <- .sanCheck(reference, startPos, indelRead, softPos, refs, alt, splitCigar)
        }
        
        # Sanity check!
        #comp <- .sanCheck(reference, startPos, indelRead, softPos, refs, alt, splitCigar)
      } # D
      
      # Store the information
      # Only have one indel
      if (i == 1) {
        mcols(newIndelRead)$indelStart <- startPos
        mcols(newIndelRead)$indelEnd <- endPos
        mcols(newIndelRead)$alt <- alt
        mcols(newIndelRead)$ref <- refs
      }
      
      # Append another indel to the first one
      # Occurs when you have multiple indels in a single read
      else {
        
        appendIndelRead <- indelRead
        mcols(appendIndelRead)$indelStart <- startPos
        mcols(appendIndelRead)$indelEnd <- endPos
        mcols(appendIndelRead)$alt <- alt
        mcols(appendIndelRead)$ref <- refs
        
        newIndelRead <- append(newIndelRead, appendIndelRead)
      }
      
    } # For
    
  } # Else 
  
  return(newIndelRead)
}

# helper fn
# Will make turn the indel information into a MVRanges
.indelToMVR <- function(indelReads, refSupport, ref, bam, covg) {
  
  # This is mostly copied and pasted from the mvr created at the end of pileup
  indelSupport <- which(!is.na(mcols(indelReads)$alt))
  indelSupport <- indelReads[indelSupport]
  indelSupport <- as.data.frame(indelSupport)
  
  indelSupport$which_label <- NULL
  
  indelSupport$refDepth <- NA_integer_
  indelSupport$altDepth <- 1
  indelSupport$totalDepth <- 1
  
  indelSupport$isAlt <- TRUE
  
  ### Not sure what this line is supposed to do
  #pu$alleles <- .byPos(pu, "isAlt") + 1 
  
  #### DO I WANT TO DO THIS??
  indelSupport$pos <- as.numeric(indelSupport$indelStart)
  
  # Turn into granges then turn into vranges
  columns <- c("seqnames","pos","ref","alt","totalDepth","refDepth","altDepth")
  grIndel <- keepSeqlevels(.puToGR(subset(indelSupport, isAlt==1)[,columns]), unique(indelSupport$seqnames))
  seqinfo(grIndel) <- Seqinfo(levels(seqnames(grIndel)), width(.getRefSeq(ref)), isCircular=TRUE, genome=ref)
  
  vrIndel <- makeVRangesFromGRanges(.puToGR(subset(indelSupport, isAlt==1)[,columns]))
  vrIndel <- keepSeqlevels(vrIndel, levels(seqnames(grIndel))) 
  seqinfo(vrIndel) <- seqinfo(grIndel)
  
  # Skipped this coverage step
  ### IS THAT ALRIGHT?
  #mvr <- MVRanges(vr, coverage=median(rowsum(pu$count, pu$pos)))
  
  # Turn into mvranges
  mvrIndel <- MVRanges(vrIndel, coverage = covg)
  sampleNames(mvrIndel) <- base::sub(paste0(".", ref), "", base::sub(".bam", "",  basename(bam)))
  
  # Fix ranges
  indelRanges <- IRanges(as.numeric(indelSupport$indelStart), as.numeric(indelSupport$indelEnd))
  ranges(mvrIndel) <- indelRanges
  
  # Assign names
  names(mvrIndel)[which(ref(mvrIndel) != alt(mvrIndel))] <- MTHGVS(subset(mvrIndel, ref != alt)) 
  
  # Only keep unique variants (based upon names that were just assigned)
  mvrIndelunique <- unique(mvrIndel)
  
  # Assign altDepths based upon what was unique
  # Sometimes the order gets mixed up, so gotta keep track of that
  order <- match(names(mvrIndelunique), names(table(names(mvrIndel))))
  count <- unname(table(names(mvrIndel))[order])
  altDepth(mvrIndelunique) <- count
  
  # Now have to get the refDepth from indelReads that were all matched
  # Added this as an argument to speed up processes
  
  #refDepth(mvrIndelunique) <- 0
  #refSupport <- which(is.na(mcols(indelReads)$alt))
  #refSupport <- indelReads[refSupport]
  
  # Need to keep order in mind, since not every indel has a read supporting the reference
  # Sets the refDepth to corresponding index that is matched
  ov <- findOverlaps(refSupport, mvrIndelunique)
  refCount <- table(subjectHits(ov))
  refDepth(mvrIndelunique[as.numeric(names(refCount))]) <- as.numeric(unname(refCount))
  
  # Replace NA's with 0s for refDepth to properly calculate totalDepth
  noRef <- which(is.na(refDepth(mvrIndelunique)))
  refDepth(mvrIndelunique[noRef]) <- 0
  
  # Recalculate totalDepth
  totalDepth(mvrIndelunique) <- refDepth(mvrIndelunique) + altDepth(mvrIndelunique)
  
  # Want to make sure we keep track of coverage
  ### NOT SURE HOW TO KEEP TRACK OF THE COVERAGE
  #mvr <- MVRanges(mvrIndelunique, coverage = median(rowsum(altDepth(mvrIndel), start(mvrIndel))))
  mvr <- MVRanges(mvrIndelunique, coverage = covg)
  return(mvr)
  
}

# helper fn
# Usually commented out unless you run into an error
# A sanity check to compare the raw read against the reference
# Prints other information like start position, cigar string, and the calculated alt and ref
.sanCheck <- function(reference, startPos, indelRead, softPos, refs, alt, splitCigar) {

  altSeq <- as.character(mcols(indelRead)$seq)
  
  # Get the corresponding ref sequence
  start <- start(indelRead) - softPos
  if (start < 0) start <- 1
  refRange <- IRanges(start, end(indelRead))
  refSeq <- as.character(unlist(extractAt(reference, refRange)))
  
  # Compare the two strings
  # Tried doing this using positions based upon cigar string but sometimes got off by one errors
  comp <- pairwiseAlignment(refSeq, altSeq)
  show(comp)
  show(splitCigar)
  print(paste("ref: ", refs))
  print(paste("alt: ", alt))
  print(paste("start: ", as.character(startPos)))

  return(comp)
}